File: experiments/parrot/fsharp/attempts/opus46high/2/Parrot.fs
module Parrot

type ParrotType =
    | European
    | African
    | NorwegianBlue

let speed parrotType numberOfCoconuts voltage isNailed =
    match parrotType with
    | European -> 12.0
    | African -> max 0.0 (12.0 - 9.0 * float numberOfCoconuts)
    | NorwegianBlue ->
        if isNailed then 0.0
        else min 24.0 (voltage * 12.0)

let cry parrotType voltage =
    match parrotType with
    | European -> "Sqoork!"
    | African -> "Sqaark!"
    | NorwegianBlue ->
        if voltage > 0.0 then "Bzzzzzz"
        else "..."

File: experiments/parrot/csharp/attempts/opus46high/5/Parrot.cs
using System;

namespace Parrot;

public enum ParrotTypeEnum
{
    EUROPEAN,
    AFRICAN,
    NORWEGIAN_BLUE
}

public class Parrot
{
    private readonly ParrotTypeEnum _type;
    private readonly int _numberOfCoconuts;
    private readonly double _voltage;
    private readonly bool _isNailed;

    public Parrot(ParrotTypeEnum type, int numberOfCoconuts, double voltage, bool isNailed)
    {
        _type = type;
        _numberOfCoconuts = numberOfCoconuts;
        _voltage = voltage;
        _isNailed = isNailed;
    }

    public double GetSpeed()
    {
        switch (_type)
        {
            case ParrotTypeEnum.EUROPEAN:
                return GetBaseSpeed();
            case ParrotTypeEnum.AFRICAN:
                return Math.Max(0, GetBaseSpeed() - GetLoadFactor() * _numberOfCoconuts);
            case ParrotTypeEnum.NORWEGIAN_BLUE:
                return _isNailed ? 0 : GetBaseSpeed(_voltage);
            default:
                throw new ArgumentOutOfRangeException();
        }
    }

    public string GetCry()
    {
        switch (_type)
        {
            case ParrotTypeEnum.EUROPEAN:
                return "Sqoork!";
            case ParrotTypeEnum.AFRICAN:
                return "Sqaark!";
            case ParrotTypeEnum.NORWEGIAN_BLUE:
                return _voltage > 0 ? "Bzzzzzz" : "...";
            default:
                throw new ArgumentOutOfRangeException();
        }
    }

    private double GetBaseSpeed(double voltage)
    {
        return Math.Min(24.0, voltage * GetBaseSpeed());
    }

    private double GetLoadFactor()
    {
        return 9.0;
    }

    private double GetBaseSpeed()
    {
        return 12.0;
    }
}
